### DISCUSSION TITLE
sequencing rules with Action onSuccess and onFailure : how to see the details of the rule result?

### DISCUSSION THREAD

**Discussion by gedoFR** [microsoft/RulesEngine#118] [Q&A]
Hello 

I have configured the workflow to used the Actions onFailure and onSuccess. 

`[
  {
    "WorkflowName": "inputWorkflowDECO",
    "Rules": [
      {
        "RuleName": "checkFlightNumberChaining",
        "SuccessEvent": "10",
        "ErrorMessage": "checkFlightNumberChaining FAILED",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "flight.carrier != null ",
        "Actions": {
          "onSuccess": {
            "Name": "EvaluateRule",
            "Context": {
              "workflowName": "nestedWorkflow",
              "ruleName": "checkSegmentOnSuccess"
            }
          }
        }
      }
    ]
  },
  {
    "WorkflowName": "nestedWorkflow",
    "Rules": [
      {
        "RuleName": "checkSegmentOnSuccess",
        "SuccessEvent": "20",
        "ErrorMessage": "checkSegmentOnSuccess failed ",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "segment1Start.airport == \"ERROR\" ",
        "Actions": {
          "onFailure": {
            "Name": "EvaluateRule",
            "Context": {
              "workflowName": "nestedWorkflow",
              "ruleName": "checkSegmentOnFailure"
            }
          }
        }
      },
      {
        "RuleName": "checkSegmentOnFailure",
        "SuccessEvent": "30",
        "ErrorMessage": "One or more adjust rules failed. ",
        "ErrorType": "Error",
        "RuleExpressionType": "LambdaExpression",
        "Expression": "segment1End.airport != null ",
        "Actions": {
          "onSuccess": {
            "Name": "OutputExpression",
            "Context": {
              "expression": "FINAL chaining rule: outputExpression on success"
            }
          }
        }
      }
    ]
  }
]`

I wanted to chain the rules together , to have a final output action and follow what was going on during the process evaluation. 

The first rule on success calls a second rule coming from another workflowName which on failure calls a third rule. 

As final result, i've got an expected IsSuccess , but I'm not able to follow what is happening during the different rule evaluation.
The onSuccess/OutputExpression from the latest rule  is not displayed somewhere.

I wanted to add the ruleResultTree in the result.ActionResult 
For that I have changed the RuleEngine.cs code like this : 
![image](https://user-images.githubusercontent.com/82643523/115014556-ca47c300-9eb2-11eb-9077-77ff7075ccb1.png)

`     public async ValueTask<List<RuleResultTree>> ExecuteAllRulesAsync(string workflowName, params RuleParameter[] ruleParams)
        {
            var ruleResultList = ValidateWorkflowAndExecuteRule(workflowName, ruleParams);
            foreach (var ruleResult in ruleResultList)
            {
                var actionResult = await ExecuteActionForRuleResult(ruleResult, true);  // gedo  set to true
                ruleResult.ActionResult = new ActionRuleResult {    // gedo changed to ActionRuleResult
                    Output = actionResult.Output,
                    Exception = actionResult.Exception,
                    Results = actionResult.Results              // gedo added Results
                };
            }
            return ruleResultList;`

**Is that the right way for having the details of each ruleAction?**

But also in that case, if a rule is unsuccess (Rule name = checkSegmentOnSuccess) 
-the dedicated errormessage ("checkSegmentOnSuccess failed") is not displayed.

![image](https://user-images.githubusercontent.com/82643523/115019187-21e92d00-9eb9-11eb-84b8-0b27d61a6991.png)

Do you have any feedback about that topic, it that something that will be extended in the future ? 

thank you for any updates, and congratulations to the developer for this nice engine

**Comment by abbasc52** [comment]
Hi @gedoFR , 

I believe you are looking for Rule Chaining instead of the default "Run All" mode.
It is already available in RulesEngine but not documented currently to prevent confusion.

you need to use `ExecuteActionWorkflowAsync` method instead of `ExecuteAllRulesAsync`. 
It will have the final output of action and the list of ruleExecution,

**Reply by gedoFR** [reply]
First of all thanks for your answer

**Comment by gedoFR** [comment]
Hi @abbasc52 .

I made some tests with your RulesEngineWithActionsTests.cs class 
Test 1 : (missing error message when the rule fails) 
I voluntarily defined an error in the "ExpressionOutputRuleTest" rule and I ran the test[WhenExpressionIsSuccess_OutputExpressionAction_ReturnsExpressionEvaluation()]
![image](https://user-images.githubusercontent.com/82643523/115695663-471ae700-a362-11eb-8f2d-dd15c9db1b32.png)

I have this response, where I expected to get the error message in the rule.
![image](https://user-images.githubusercontent.com/82643523/115696224-c01a3e80-a362-11eb-9c88-56dfd9b61c58.png)


Test2 : (missing error message when the rule fails, even if an action onFailure is defined)
I redefined the rule "EvaluateRuleTest" with an action to evaluate the following rule onFailure and I defined an error message in the rule 
And i ran the test [WhenExpressionIsSuccess_EvaluateRuleAction_ReturnsExpressionEvaluation()]
![image](https://user-images.githubusercontent.com/82643523/115696887-5ea69f80-a363-11eb-81c3-fb59a7e3597d.png)

I have this, where I also expected to know the reason why a rule failed even though the chaining is correct, 
![image](https://user-images.githubusercontent.com/82643523/115697672-091ec280-a364-11eb-9641-72f8dbfbe3f0.png)


For my use, it is really important that I have the details of the failed rules. I won't be using chaining so much.
My most important need is for rules with nested rules.
I am also not able to see the details of the error message in this case.
I saw that a way out is open for this.
Is a new version planned on this subject, and maybe when is it planned?

**Comment by poojak0403** [comment]
Hi all,
What is the solution to the above issue? I also need to show error messages for the particular failed rules. 

![re](https://user-images.githubusercontent.com/94694237/215308702-c41b25ec-c857-4e48-af48-9b421583376a.png)
